src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,329 @@
+     * To ensure locks don't expire prematurely; the lock TTL should be set
+     * with enough extra time to account for any clock drift between nodes.
+     * @see self::createTTLIndex()
+     *
+     * writeConcern, readConcern and readPreference are not specified by
+     * MongoDbStore meaning the collection's settings will take effect.
+     * @see https://docs.mongodb.com/manual/applications/replication/
+     */
+    public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)
+    {
+        if (!isset($options['database'])) {
+            throw new InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+        ), $options);
+
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * Create a TTL index to automatically remove expired locks.
+     *
+     * This should be called once during database setup.
+     *
+     * Alternatively the TTL index can be created manually:
+     *
+     *  db.lock.ensureIndex(
+     *      { "expires_at": 1 },
+     *      { "expireAfterSeconds": 0 }
+     *  )
+     *
+     * Please note, expires_at is based on the application server. If the
+     * database time differs; a lock could be cleaned up before it has expired.
+     * Set a positive expireAfterSeconds to account for any time drift between
+     * application and database server.
+     *
+     * A TTL index MUST BE used on MongoDB 2.2+ to automatically clean up expired locks.
+     *
+     * @see http://docs.mongodb.org/manual/tutorial/expire-data/
+     *
+     * @return string The name of the created index
+     *
+     * @throws UnsupportedException          if options are not supported by the selected server
+     * @throws MongoInvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException        for other driver errors (e.g. connection errors)
+     */
+    public function createTTLIndex(): string
+    {
+        $keys = array(
+            'expires_at' => 1,
+        );
+
+        $options = array(
+            'expireAfterSeconds' => 0,
+        );
+
+        return $this->getCollection()->createIndex($keys, $options);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws LockStorageException if failed to save to storage (fallback exception)
+     * @throws LockExpiredException when save is called on an expired lock
+     */
+    public function save(Key $key)
+    {
+        $token = $this->getUniqueToken($key);
+        $now = microtime(true);
+
+        $filter = array(
+            '_id' => (string) $key,
+            '$or' => array(
+                array(
+                    'token' => $token,
+                ),
+                array(
+                    'expires_at' => array(
+                        '$lte' => $this->createDateTime($now),
```

Comment:
1. why do you need the `$lte`
2. why not `$this->createDateTime($now + $this->initialTtl)` instead?

Reply:
Sorry, I don't really follow... This is a filter looking for locks that have expired.

Reply:
```sql
UPDATE OR INSERT lock
SET
    id = :key,
    token = :token
    expires_at = :newExpiry
WHERE
    id = :key
    AND (
        token = :key
        OR expires_at <= :now
    )
```
Does this explain it? (Assuming `UPDATE OR INSERT` is a real thing and it's ACID compliant being 1 statement).

Reply:
The logic makes more sense now. Thanks for explaining.